sound/soc/sof/intel/hda.c

```diff
@@ -41,6 +41,62 @@
 #define EXCEPT_MAX_HDR_SIZE	0x400
 #define HDA_EXT_ROM_STATUS_SIZE 8
 
+/*
+ * Helper function to set up a DAI widget in the DSP and then send the updated DAI config during
+ * hw_params or send the updated DAI_CONFIG and then free the DAI widget during hw_free
+ */
+int hda_dai_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *sof_dai,
+			 struct sof_ipc_dai_config *config, bool setup, int dir)
+{
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_reply reply;
+	bool found = false;
+	int ret;
+
+	/* find snd_sof_widget of type DAI with matching sname as that of snd_sof_dai name */
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (dir == SNDRV_PCM_STREAM_PLAYBACK && swidget->id != snd_soc_dapm_dai_in)
+			continue;
+
+		if (dir == SNDRV_PCM_STREAM_CAPTURE && swidget->id != snd_soc_dapm_dai_out)
+			continue;
+
+		if (!strcmp(sof_dai->name, swidget->widget->sname)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		dev_err(sdev->dev, "error: failed to find DAI widget for: %s\n",
+			sof_dai->cpu_dai_name);
+		return -EINVAL;
+	}
+
+	/*
+	 * For static pipelines, the DAI widget would already be set up and calling
+	 * sof_widget_setup()/free() would simply return without doing anything.
+	 * For dynamic pipelines, the DAI widget would be set up or freed here.
```

Comment:
The conditional "would be set up here" is a bit weird. I mean this reads as a placeholder for the actual code. The actions are already implemented, so "For dynamic pipelines, the DAI widget is set up or freed here." would be more accurate, right.